### PR TITLE
net: tc: Priority to Traffic Class mapping improvements

### PR DESCRIPTION
**Pull request by tgorochowik**
This commits adds new priority to traffic class mappings and allows
users to choose which mapping to use through menuconfig.

The new mappings are recommended in 802.1 (chapter 33.5) for
time-sensitive applications supporting the credit-based sharper
algorithm.

Please review, thanks!

### CODE REVIEW

**Review by jukkar — CHANGES_REQUESTED** [review]
Minor tweak needed, otherwise ok.

**Comment by jukkar** [comment]
We also need to test these new mappings. Different combinations of the new mappings need to be added to tests/net/traffic_class/testcase.yaml file. Not sure if the actual unit test would work as is or does it need modifications.

File: subsys/net/ip/net_tc_mapping.h [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+#ifndef __TC_MAPPING_H
+#define __TC_MAPPING_H
+
+#include <zephyr/types.h>
+
+/* All the maps below use priorities and indexes, below is the list of them
+ * according to 802.1Q - table I-2.
+ *
+ *  Priority         Acronym   Traffic types
+ *  1 (lowest)       BK        Background
```
Comment (jukkar) [inline subsys/net/ip/net_tc_mapping.h]: Please fix also the NET_PRIORITY_xx settings in net_ip.h

**Comment by tgorochowik** [comment]
@jukkar Thank you for your review!

The maps maps have been changed a little bit as in the previous implementation it did not take into account that there could be a different number of TX and RX TC classes. Is the current implementation ok? With it there is no need to duplicate the maps for RX and TX.

The issues you raised in your review are addressed - there is just one thing to verify - I am not sure if the PCP (vlan priority) to Packet priority conversion is still needed.

It looks like both the PCP values and Packet Priorities values are swapped (BE and BK priorities). So earlier we did fix VLAN priorities conversion, but it seems that not in a proper way. This standard is just so confusing when it comes to priorities.

File: subsys/net/ip/net_tc_mapping.h (hunk 1 above)
Comment (tgorochowik) [reply subsys/net/ip/net_tc_mapping.h]: Added a commit that does just that.

File: subsys/net/ip/net_tc.c [hunk 2]
```diff
@@ -52,8 +52,8 @@ int net_tx_priority2tc(enum net_priority prio)
 	 * chapter 34.5 table 34-1
```
Comment (jukkar) [inline subsys/net/ip/net_tc.c]: The commit message should probably say "...but 1 is the lowest."

**Review by jukkar — APPROVED** [review]
LGTM

**Review by jukkar — CHANGES_REQUESTED** [review]
Forgot that the commit message needs still fixing.

**Review by jukkar — APPROVED** [review]

File: subsys/net/ip/Kconfig [hunk 3]
```diff
@@ -70,6 +70,39 @@ config NET_TC_RX_COUNT
...
+	help
+	  Select mapping to use to map network packet priorities to traffic
+	  classes.
+
+config NET_TC_MAPPING_STRICT
+	bool "Strict priority mapping"
+	help
+	  This is the recommended default priority to traffic class mapping.
+	  Use it for implementations that do not support the credit-based
+	  sharper transmission selection algorithm.
```
Comment (pfalcon) [inline subsys/net/ip/Kconfig]: @tgorochowik: Post-review: a typo sneaked in, s/sharper/shaper/. Hope you can see this and submit a patch ;-).

File: subsys/net/ip/net_tc_mapping.h [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+ */
+
+/* Helper macros used to generate the map to use */
+#define PRIORITY2TC_GEN_INNER(TYPE, COUNT) priority2tc_ ## TYPE ## _ ## COUNT
+#define PRIORITY2TC_GEN(TYPE, COUNT) PRIORITY2TC_GEN_INNER(TYPE, COUNT)
+
+#if defined(CONFIG_NET_TC_MAPPING_STRICT)
+
+/* This is the recommended priority to traffic class mapping for
+ * implementations that do not support the credit-based sharper transmission
```
Comment (pfalcon) [inline subsys/net/ip/net_tc_mapping.h]: Also "sharper". Probably worth doing search-and-replace...

Comment (tgorochowik) [reply subsys/net/ip/net_tc_mapping.h]: @pfalcon Right! Sorry about that! This is actually a copy of the same text (Header and Kconfig), thank you for letting me know :+1: I'll submit a patch for that.

**Pull request merged**